src/System.Memory/ref/System.Memory.cs

```diff
@@ -353,3 +357,10 @@ public static class Utf8Parser
     }
 }
 
+namespace System.Runtime.InteropServices
+{
+    public static class MemoryMarshal
+    {
+        public static Memory<T> AsMemory<T>(ReadOnlyMemory<T> readOnlyMemory) { throw null; }
```

Comment:
Do we need to add this to System.Runtime ref as well?

Reply:
It depends on whether we want to use it in code at a lower level than System.Memory, or whether it's ok for components to need to reference System.Memory in order to get it.  What are we doing with the other classes, like SpanExtensions (MemoryExtensions)... are those going to be in System.Runtime?

Reply:
No, they won't be in System.Runtime given they are contain extension methods only.

Reply:
> given they are contain extension methods only

Why does that matter?

Reply:
I assumed extension methods are more like opt-in as needed.

Reply:
I see. I don't think it's so much about opt-in-ness but about dependencies and layering.  Something needs to be exposed from the System.Runtime.dll contract if something else in the System.Runtime.dll contract uses it in public API, and that won't be the case for a static class.  Something would need to be exposed from lower than System.Memory.dll if something else in corefx needed to use it and couldn't depend on System.Memory.dll for some reason.  But System.Memory.dll is low enough in the layering that there should be very few assemblies where that's the case.  A larger concern for me around the extension method isn't that they're exposed from System.Memory.dll, but that they're not available for use inside corelib, such that we've had to duplicate some of the functionality in order to implement various things in corelib, and those duplicated implementations aren't as good / efficient / complex as the ones in System.Memory.dll.

Reply:
> I don't think it's so much about opt-in-ness but about dependencies and layering. 

That makes sense. 

> Something needs to be exposed from the System.Runtime.dll contract if something else in the System.Runtime.dll contract uses it in public API, and that won't be the case for a static class.

Why won't that be the case specifically for a static class (in this case SpanExtensions)?

> Something would need to be exposed from lower than System.Memory.dll if something else in corefx needed to use it and couldn't depend on System.Memory.dll for some reason. But System.Memory.dll is low enough in the layering that there should be very few assemblies where that's the case. 

Is that why the assemblies in corefx that now use Span reference System.Memory.dll and not just System.Runtime? So that they have access to the extension methods like string.AsReadOnlySpan() or NonPortableCast?

I was able to find three assemblies that reference System.Memory atm:
- [System.Net.Primitives](https://github.com/dotnet/corefx/blob/master/src/System.Net.Primitives/src/System.Net.Primitives.csproj) and [ref](https://github.com/dotnet/corefx/blob/master/src/System.Net.Primitives/ref/System.Net.Primitives.csproj)
- [System.Net.Security](https://github.com/dotnet/corefx/blob/master/src/System.Net.Security/src/System.Net.Security.csproj) and [ref](https://github.com/dotnet/corefx/blob/master/src/System.Net.Security/ref/System.Net.Security.csproj)
- [System.Net.WebSockets](https://github.com/dotnet/corefx/blob/master/src/System.Net.WebSockets/src/System.Net.WebSockets.csproj) (not in ref since no APIs expose Span/Memory).

> A larger concern for me around the extension method isn't that they're exposed from System.Memory.dll, but that they're not available for use inside corelib, such that we've had to duplicate some of the functionality in order to implement various things in corelib, and those duplicated implementations aren't as good / efficient / complex as the ones in System.Memory.dll.

Can you share an example of a sub-par duplicated implementation?

Reply:
> Why won't that be the case specifically for a static class (in this case, SpanExtensions)?

When I said "uses it in public API", I literally meant in the API signature.  A static class can't have instances, and thus it wouldn't be in a signature.

> So that they have access to the extension methods like string.AsReadOnlySpan(), NonPortableCast, or SequenceEqual (some of the extension methods I saw being used)?

Yes, for AsReadOnlySpan, for IndexOf, etc.  If we don't want such assemblies to reference System.Memory, SpanExtensions will need to be moved down.

> I was able to find three assemblies that reference System.Memory atm

Also System.Net.Http, System.IO.FileSystem, and System.Data.SqlClient.  And that's going to continue to grow, because, for example, AsReadOnlySpan is the only way to use a string as a span.

> Can you share an example of a sub-par duplicated implementation?

https://github.com/dotnet/coreclr/blob/master/src/mscorlib/shared/System/StringSpanHelpers.cs, for example the trivial IndexOf implementation here versus the optimized implementation in https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System/SpanExtensions.cs.

Reply:
>https://github.com/dotnet/coreclr/blob/master/src/mscorlib/shared/System/StringSpanHelpers.cs, for example the trivial IndexOf implementation here versus the optimized implementation in https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System/SpanExtensions.cs

This is a good reason why to move SpanExtensions into CoreLib. @ahsonkhan Could you please add it to your backlog?

Reply:
Opened https://github.com/dotnet/corefx/issues/25182